### PR TITLE
Implement descriptions

### PR DESCRIPTION
**Pull request by SimonBaeumer**
This Pull-Requests adds the option to add descriptions into the sh scripts, i.e.
```
#!/usr/bin/env bash
#DESCRIPTION: My funny description.
```
And will display something like this:
```
###################
Available commands:

 - build    Builds phar packages
 - clear    Remove build artifacts
 - unit     Execute unit tests

3 script(s) available
```

@JanPietrzyk Humbug fails maybe due dependencies which are used by the project and humbug's phar, if I add humbug to composer directly I have no problems with executing it :/
See https://github.com/humbug/humbug/issues/39

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -185,6 +185,15 @@ header: |
                     |_|
 ```
 
+#### DESCRIPTION
+
+You can add a description to a script which will be printed when the command list will be displayed.
+
+```sh
+#!/usr/bin/env bash
+#DESCRIPTION: My usefull comment.
```
Comment (bcremer) [inline README.md]: Typo: usefull -> useful

File: src/Listing/DescriptionReader.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+class DescriptionReader
+{
+    const KEY_WORD = 'DESCRIPTION:';
+
+    /**
+     * @param array $lines
+     * @return string
+     */
+    public function read(array $lines)
```
Comment (SimonBaeumer) [inline src/Listing/DescriptionReader.php]: self cr, type hint fehlt

**Comment by JanPietrzyk** [comment]
Regarding the build: You usage of PHPUnits mock builder seems to be the problem. I removed all calls locally and Humbug works again.

File: src/Listing/ScriptFinder.php [hunk 3]
```diff
@@ -52,7 +58,9 @@ public function getAllScripts(): array
                     continue;
                 }
 
-                $newScript = new Script($path->getPath(), $fileName, $path->getNamespace());
+                $description = $this->scriptDescriptionReader->read(file($path->getPath() . '/' . $fileName));
```
Comment (JanPietrzyk) [inline src/Listing/ScriptFinder.php]: why add a description reader that then is not reading the file content?

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 4]
```diff
@@ -5,36 +5,53 @@
...
 {
+    /**
+     * @var DescriptionReader|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $descriptionReaderMock;
+
     public function test_script_finder_holds_contract_if_no_paths_present()
     {
-        $finder = new ScriptFinder([]);
+        $finder = new ScriptFinder([], $this->createMock(DescriptionReader::class));
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: the real one could be used as easily here

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 5]
```diff
@@ -5,36 +5,53 @@
...
-        $this->assertCount(3, $finder->getAllScripts());
+        $this->assertCount(4, $finder->getAllScripts());
     }
 
     public function test_script_finder_finds_scripts_if_two_directories_are_passed_and_filters_noise()
     {
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')],
+            $this->descriptionReaderMock
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: the real one could be used as easily here

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 6]
```diff
@@ -5,36 +5,53 @@
...
+        $this->assertCount(4, $finder->getAllScripts());
         $this->assertContainsOnlyInstancesOf(Script::class, $finder->getAllScripts());
     }
 
     public function test_script_finder_finds_script_by_name_if_two_directories_are_passed_and_filters_noise()
     {
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')],
+            $this->descriptionReaderMock
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: the real one could be used as easily here

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 7]
```diff
@@ -5,36 +5,53 @@
...
         $this->assertInstanceOf(ScriptFinder::class, $finder);
         $this->assertInternalType('array', $finder->getAllScripts());
     }
 
     public function test_script_finder_finds_scripts_if_one_directory_is_passed()
     {
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts')],
+            $this->descriptionReaderMock
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: the real one could be used as easily here

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 8]
```diff
@@ -44,7 +61,10 @@ public function test_script_finder_finds_script_by_name_if_two_directories_are_p
 
     public function test_script_finder_prefixes_script_names_with_namespace_if_present()
     {
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff', 'biz')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff', 'biz')],
+            $this->descriptionReaderMock
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: the real one could be used as easily here

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 9]
```diff
@@ -54,8 +74,31 @@ public function test_script_finder_prefixes_script_names_with_namespace_if_prese
 
     public function test_script_finder_throws_exception_if_path_is_not_valid()
     {
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts_not_valid_directory')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts_not_valid_directory')],
+            $this->descriptionReaderMock
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: the real one could be used as easily here

File: tests/Integration/Listing/ScriptFinderTest.php [hunk 10]
```diff
@@ -54,8 +74,31 @@ public function test_script_finder_prefixes_script_names_with_namespace_if_prese
...
+        );
+
+        $script = $finder->findScriptByName('description');
+        $this->assertSame('My description', $script->getDescription());
+    }
+
+    /**
+     * @before
+     */
+    public function createEmptyDescriptionReaderMock()
```
Comment (JanPietrzyk) [inline tests/Integration/Listing/ScriptFinderTest.php]: please remove. And if you need a method like that I would prefer a setUp method on top of the test file

**Review by JanPietrzyk — CHANGES_REQUESTED** [review]
a few last tweaks please

**Comment by SimonBaeumer** [comment]
@JanPietrzyk I removed `createMock` calls and moved the file reading to the `DescriptionReader` class. I didn't have a reason why I implemented the file reading in the `ScriptFinder` ^^

**Pull request merged**